### PR TITLE
Support for additional JWT signing methods

### PR DESCRIPTION
**Pull request by yannh**
This is a follow-up to https://github.com/vouch/vouch-proxy/issues/327  and a work-in-progress - I've opened this to get early feedback.

This introduces the ability to sign vouch's JWT using asymmetric encryption algorithms, so that the site proxied by nginx+vouch can verify the JWT using a public key, and verifying the claims from the JWT before granting access, ensuring the request does not come from an unauthenticated user bypassing the Nginx proxy.

Note that this is very early code, and needs additional tests. I would especially appreciate feedback on:
* the changes to the configuration file
* the approach to test using multiple config files :thinking: 
* and more generally whether this feature is worth pursuing.

Note1: I just looked up how Pomerium does this https://www.pomerium.com/docs/topics/getting-users-identity.html#prerequisites  - it seems they sign the JWT using ES256 - so this PR would enable a similar behaviour.

Note2: the keys are generated using the following commands- maybe it would be nicer to make a target in do.sh to generate them? :thinking:  EDIT: As I thought, received an email from Github that this has been detected by "gitguardian" as a security risk. I've deleted them for now, but unit tests will fail.
openssl genrsa -out config/testing/rsa.key 4096
openssl rsa -in config/testing/rsa.key -pubout > config/testing/rsa.pub

### CODE REVIEW

File: pkg/jwtmanager/jwtmanager.go [hunk 1]
```diff
@@ -73,6 +74,72 @@ func populateSites() {
...
 }
 
+func decryptionKey() (interface{}, error) {
+	if strings.HasPrefix(cfg.Cfg.JWT.SigningMethod, "HS") {
+		return []byte(cfg.Cfg.JWT.Secret), nil
+	}
+
+	f, err := os.Open(cfg.Cfg.JWT.PublicKeyFile)
+	if err != nil {
+		return nil, fmt.Errorf("error opening Key %s: %s\n", cfg.Cfg.JWT.PrivateKeyFile, err)
```
Comment (bnfinet) [inline pkg/jwtmanager/jwtmanager.go:84]: the error is for the PublicKeyFile but the error message is for the PrivateKeyFil

**Comment by yannh** [comment]
I ve added the keys back with a file to indicate to gitguardian that the test keys are intended to be here. Tests now pass. I have added more validation in the config parsing. I would love to add more tests around config file reading, but that might be quite a significant refactor. The only documentation for this new feature is in config/config.yml_example - is this enough?

**Comment by bnfinet** [comment]
@yannh my apologies for the delays in reviewing this code.

Overall it LGTM.  There are a few improvements which I'd like to have before merge...

* move crypto config including key parsing from  `jwtmanager.go` to `cfg.go` or perhaps `jwtmanager.go Configure()`
* instead of publishing keys to this repo, generate keys during testing by calling a method in `./do.sh` as you suggested

Thanks again for the contribution.  This will be an excellent addition to VP.

**Comment by yannh** [comment]
Thanks a lot for the review @bnfinet ! I will finalize the PR soon. :rocket:

**Comment by bnfinet** [comment]
@yannh hurry hurry!  2021 is 1/6th of the way over!!

Do let me know if there's anything that I can do to help move this forward. :)

**Comment by yannh** [comment]
Yes! I'm still here, hoping to get to it soon :)

**Comment by yannh** [comment]
@bnfinet aaand tests pass again :) Couple of notes... Generating  the certs in do.sh introduces a dep on openssl, I sudo apt-get install it in travis, but for people who run tests locally, you need to keep that in mind...
Also - I hit the same issue as this https://github.com/golang-migrate/migrate/issues/480  with go get, I updated go to 1.16 which fixed it.

**Comment by bnfinet** [comment]
@yannh thank you for the fine addition to Vouch Proxy.  Very much appreciated.

Here's looking forward to other users benefiting from your work in adding RSA and other signing mechanisms to VP.

If you ever find yourself on the US west coast please consider stopping by Portland.  I'd love to buy you a beverage of your choice.

**Pull request merged**